lib/application.js

```diff
@@ -241,6 +244,66 @@ app.use = function use(fn) {
 };
 
 /**
+ * Proxy `Router#disuse()` to remove middleware from the app router.
+ * See Router#disuse() documentation for details.
+ *
+ * If the _fn_ parameter is an express app, then it will be
+ * dismounted from the _route_ specified.
+ *
+ * @public
+ */
+
+app.disuse = function use(fn) {
+  var offset = 0;
+  var path = '/';
+
+  // default path to '/'
+  // disambiguate app.use([fn])
+  if (typeof fn !== 'function')  {
+    var arg = fn;
+
+    while (Array.isArray(arg) && arg.length !== 0)  {
+      arg = arg[0];
+    }
+
+    // first arg is the path
+    if (typeof arg !== 'function')  {
+      offset = 1;
+      path = fn;
+    }
+  }
+
+  var fns = flatten(slice.call(arguments, offset));
+
+  if (fns.length === 0)  {
+    throw new TypeError('app.use() requires middleware functions');
+  }
+
+  // setup router
+  this.lazyrouter();
+  var router = this._router;
+
+  fns.forEach(function (fn)  {
+    // non-express app
+    if (!fn || !fn.handle || !fn.set)  {
+      return router.disuse(path, fn);
+    }
+
+    debug('.disuse app under %s', path);
```

Comment:
Cool. Only two real issue right now:

1. The stack needs to get consolidated no matter what. Right now, there has to be no active requests, which the only way to even guarantee that is for the app to stop taking requests altogether. This should not be a requirement in order to clean up the stack; it should clean up even if there is never a time when a request is not being processed. The reason is that if you have to stop requests from coming in to get it cleaned up, at that point, you could have simply cycled the worker process altogether and rebuilt the entire stack without `disuse`.
2. Still need those benchmarks for the router changes.